### PR TITLE
2103 Add deepcopy to PosNegCrop transform

### LINKED ISSUE #2103: Need to deepcopy data in RandCropByPosNegLabeld transform
**Describe the bug**
Currently, when we crop samples in `RandCropByPosNegLabeld` transform, it uses shallow copy for meta_data and other information, then if we change some value in a sample, all the samples are changed.
Need to use `deepcopy` for it, similar issue to the `RandSpatialCropSamplesd`.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #2103 .

### Description
This PR changed to use `deepcopy` and added more unit tests.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: tests/test_rand_weighted_cropd.py [hunk 1]
```diff
@@ -139,6 +139,24 @@ def test_rand_weighted_crop_bad_w(self):
...
+        crop = RandWeightedCropd(("img", "seg"), "w", (10, -1, -1), n_samples)
+        weight = np.zeros_like(img)
+        weight[0, 7, 17] = 1.1
+        weight[0, 13, 31] = 1.1
+        weight[0, 24, 21] = 1
+        crop.set_random_state(10)
+        result = crop({"img": img, "seg": self.segn[0], "w": weight, "img_meta_dict": {"affine": None}})
+        self.assertTrue(len(result) == n_samples)
+        np.testing.assert_allclose(np.asarray(crop.centers), [[14, 32, 40], [41, 32, 40], [20, 32, 40]])
+        for i in range(n_samples):
```
Comment (rijobro) [inline tests/test_rand_weighted_cropd.py:154]: should the unit test not try to modify some metadata of one of the samples, and then verifying that it hasn't changed for the other samples? (check copy was deep, not shallow)

**Review by rijobro — APPROVED** [review]

File: tests/test_rand_weighted_cropd.py (hunk 1 above)
Comment (Nic-Ma) [reply tests/test_rand_weighted_cropd.py:154]: Hi @rijobro ,

I already added the `"img_meta_dict": {"affine": None}` in the test data, if shallow copy, all the samples from `img` will have the same patch index.

Thanks.

**Pull request merged**